// dotvvm-jscomponent-react/dist/dotvvm-react.js
import './node_modules/react/index.js';
import './node_modules/react-dom/index.js';
import { r as react } from './_virtual/index.js_commonjs-module';
import { r as reactDom } from './_virtual/index2.js_commonjs-module';

/// <reference path="../../Framework/Framework/obj/typescript-types/dotvvm.d.ts" />
class KnockoutTemplateReactComponent extends react.exports.Component {
    constructor() {
        super(...arguments);
        this.wrapRef = react.exports.createRef();
        this.templateName = ko.observable();
    }
    // TODO: how to dispose the template?
    // componentWillUnmount() {
    // }
    componentDidMount() {
        setTimeout(() => this.initializeTemplate(), 5);
    }
    initializeTemplate() {
        const e = this.wrapRef.current;
        let context = ko.contextFor(e);
        if (this.props.getChildContext) {
            context = this.props.getChildContext(context);
        }
        else if (this.props.viewModel !== undefined) {
            const updateEvent = new dotvvm.DotvvmEvent("templateInReact.newState");
            this.viewModelStateManager = new dotvvm.StateManager(this.props.viewModel, updateEvent);
            context = context.createChildContext(this.viewModelStateManager.stateObservable);
        }
        this.updateStuff();
        ko.renderTemplate(this.templateName, context, {}, e);
    }
    componentDidUpdate() {
        this.updateStuff();
    }
    updateStuff() {
        if (this.templateName() !== this.props.templateName)
            this.templateName(this.props.templateName);
        if (this.viewModelStateManager) {
            this.viewModelStateManager.setState(this.props.viewModel);
        }
    }
    render() {
        return react.exports.createElement(this.props.wrapperTag, { ref: this.wrapRef });
    }
}
KnockoutTemplateReactComponent.defaultProps = {
    wrapperTag: "div"
};
const registerReactControl = (ReactControl, defaultProps = {}) => ({
    create: (elm, props, commands, templates) => {
        const initialProps = Object.assign(Object.assign(Object.assign({}, defaultProps), commands), templates);
        let currentProps = Object.assign(Object.assign({}, initialProps), props);
        reactDom.exports.render(react.exports.createElement(ReactControl, Object.assign({}, currentProps)), elm);
        return {
            updateProps(updatedProps) {
                currentProps = Object.assign(Object.assign({}, currentProps), updatedProps);
                reactDom.exports.render(react.exports.createElement(ReactControl, Object.assign({}, currentProps)), elm);
            },
            dispose() {
                reactDom.exports.unmountComponentAtNode(elm);
            }
        };
    }
});

export { KnockoutTemplateReactComponent, registerReactControl };
